ultralytics/data/dataset_manager.py

```diff
@@ -0,0 +1,313 @@
+            color=colors,
+        )
+        plt.xlabel("Classes")
+        plt.ylabel("Number of Images")
+        plt.title("Classes Distribution Map")
+
+        plt.xticks(range(len(class_counts)), self.class_names, rotation=45, ha="right")
+
+        plt.tight_layout()
+        plt.show()
+
+    def get_class_fraction_images(self, class_fraction):
+        class_counts = [len(images) for images in self.images_by_classes]
+        class_counts_needed = [
+            max(round(count * class_fraction), 1) for count in class_counts
+        ]
+
+        selected_images = []
+        for cls, images in enumerate(self.images_by_classes):
+            for image in images:
+                if class_counts_needed[cls] > 0:
+                    added = True
+                    for cls2 in self.classes_by_images[image]:
+                        if class_counts_needed[cls2] <= 0:
+                            added = False
+                        class_counts_needed[cls2] -= 1
+                    if not added:
+                        for cls2 in self.classes_by_images[image]:
+                            class_counts_needed[cls2] += 1
+                    else:
+                        selected_images.append(image)
+                else:
+                    break
+            while class_counts_needed[cls] > 0:
+                random_img = random.choice(self.images_by_classes[cls])
+                for cls2 in self.classes_by_images[random_img]:
+                    class_counts_needed[cls2] -= 1
+                selected_images.append(random_img)
+
+        return selected_images
+
+    def dataset_balancing(self):
+        images_to_oversample = self.get_images_to_oversample()
+        self._transform_images_balance(images_to_oversample, 0)
+
+    def get_images_to_oversample(self):
+        # needed instance counts from each class for oversampling
+        class_counts = [
+            len(images_by_class) for images_by_class in self.images_by_classes
+        ]
+        max_class_count = max(class_counts)
+        class_counts_needed = [
+            max_class_count - count if count != 0 else 0 for count in class_counts
+        ]
+
+        images_to_oversample = (
+            []
+        )  # array to store image names which are going to be oversampled
+
+        # oversampling (main loop)
+        for class_ in range(len(class_counts_needed)):
+            current_class_images = self.images_by_classes[class_].copy()
+            improved = True
+            while class_counts_needed[class_] > 0 and improved:
+                improved = False
+                random.shuffle(current_class_images)  # increase randomity
+                for img in current_class_images:
+                    add = True
+                    for class2_ in self.classes_by_images[img]:
+                        if class_counts_needed[class2_] <= 0:
+                            add = False
+                            break
+                    if add:
```

Comment:
nit: for determening `add` all() can be used here instead of the inner for loop

```suggestion
                    if all(class_count_needed > 0 for class_count_needed in class_counts_needed):
```

Reply:
@vbulcsu but in the inner for loop, I only check the counts from the classes connected to the current image (not all), so the suggested line wouldn't be equivalent